stern/file_tail.go

```diff
@@ -0,0 +1,115 @@
+package stern
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+	"text/template"
+
+	"github.com/fatih/color"
+)
+
+type FileTail struct {
+	Options *TailOptions
+	tmpl    *template.Template
+	inName  string
+	in      io.Reader
+	out     io.Writer
+	errOut  io.Writer
+}
+
+func NewFileTail(tmpl *template.Template, inName string, in io.Reader, out, errOut io.Writer, options *TailOptions) *FileTail {
+	return &FileTail{
+		Options: options,
+		tmpl:    tmpl,
+		inName:  inName,
+		in:      in,
+		out:     out,
+		errOut:  errOut,
+	}
+}
+
+var fileColor *color.Color = color.New(color.FgYellow)
+
+// Start starts tailing
+func (t *FileTail) Start() error {
+	t.printStarting()
+
+	reader := bufio.NewReader(t.in)
+	err := t.ConsumeReader(reader)
+
+	return err
+}
+
+// Close stops tailing
+func (t *FileTail) Close() {
+	t.printStopping()
+}
+
+func (t *FileTail) printStarting() {
+	if !t.Options.OnlyLogLines {
+		g := color.New(color.FgHiGreen, color.Bold).SprintFunc()
+		y := fileColor.SprintFunc()
+		fmt.Fprintf(t.errOut, "%s › %s\n", g("+"), y(t.inName))
+	}
+}
+
+func (t *FileTail) printStopping() {
+	if !t.Options.OnlyLogLines {
+		r := color.New(color.FgHiRed, color.Bold).SprintFunc()
+		y := fileColor.SprintFunc()
+		fmt.Fprintf(t.errOut, "%s › %s\n", r("-"), y(t.inName))
+	}
+}
+
+// ConsumeReader reads the data from the reader and writes into the out
+// writer.
+func (t *FileTail) ConsumeReader(reader *bufio.Reader) error {
+	for {
+		line, err := reader.ReadBytes('\n')
+		if len(line) != 0 {
+			t.consumeLine(strings.TrimSuffix(string(line), "\n"))
+		}
+
+		if err != nil {
+			if err != io.EOF {
+				return err
+			}
+			return nil
+		}
+	}
+}
+
+// Print prints a color coded log message
+func (t *FileTail) Print(msg string) {
+	vm := Log{
+		Message:        msg,
+		NodeName:       "",
+		Namespace:      "",
+		PodName:        "",
+		ContainerName:  "",
+		PodColor:       color.New(color.Reset),
+		ContainerColor: color.New(color.Reset),
+	}
```

Comment:
The way I would probably use `--stdin` is to have my go-to alias/wrapper with a template e.g
```sh
s () {
  ./dist/stern --template='{{color .ContainerColor .ContainerName}} {{with $d := .Message | tryParseJSON}}{{toRFC3339Nano $d.timestamp}} {{$d.message}}{{end}}{{"\n"}}' "$@"
}
```
and when I have a log file I can just do this (because I'm lazy and don't maintain separate templates/aliases)
```sh
s --stdin < log
```
and the output will include the empty container plus an extra space. If I don't want the space I could do `{{if .ContainerName}}{{color .ContainerColor .ContainerName}} {{end}}`.

If placeholders are used it would always print the `<container>` at the start, which isn't too bad but not something I imagine any user needs. I guess it comes down to if `--stdin` should be geared towards testing or end users.

I prefer the default to be empty strings, but I can see the benefit for your case too.

Reply:
Thank you for explaining your use case in detail. In that case, it seems reasonable to leave values empty by default. If necessary, we can also consider adding a new flag such as `--placeholder` for testing later on.